### ISSUE TITLE
Phases in action list don't have to match phases defined in metavariables

### ISSUE THREAD

**Issue by knelli2** [sxs-collaboration/spectre#4082]
# Bug reports:

### Expected behavior:

If I define an action list, a build error should occur if I add a phase to the action list that isn't defined in the metavariables `default_phase_order`

### Current behavior:

With the BBH exec I had defined

```cpp
static constexpr std::array<Parallel::Phase, 8> default_phase_order{
      {Parallel::Phase::Initialization,
       Parallel::Phase::RegisterWithElementDataReader,
       Parallel::Phase::ImportInitialData,
       Parallel::Phase::InitializeInitialDataDependentQuantities,
       Parallel::Phase::InitializeTimeStepperHistory,
       Parallel::Phase::RegisterWithObserver, Parallel::Phase::Evolve,
       Parallel::Phase::Exit}};
```

but in the DG action list I accidentally defined a phase like 

```cpp
Parallel::PhaseActions<Parallel::Phase::Register, tmpl::list<dg_registration_list, Parallel::Actions::TerminatePhase>>
```

This built successfully, but during runtime I got errors that elements were trying to send data to the observer, but weren't registered (because we never ran the `Register` phase, we only ran the `RegisterWithObserver` phase).

I think there should be a compile time check for this that every phase in every action list for every component is actually in the phases we are using for an executable.

**Comment by kidder** [comment]
Well the `default_phase_order` is not necessarily equivalent to a list of all possible phases that an executable may use once you use `PhaseControl` to select the next phase...

**Comment by knelli2** [comment]
Oh. So is it a feature then to be able to put any phase I want into a component even if it's not defined in `default_phase_order`?

**Comment by kidder** [comment]
yes; for instance `LoadBalancing` and `WriteCheckpoint` are currently not in the `default_phase_order` (They may not be in an action list either).  But I can foresee wanting to switch to an AMR phase via PhaseControl which could then have an action list to execute...

I agree that it would be nice to catch the error you had at compile-time rather than run-time; I'll think about some...

I was planning to do some factoring of metavariables and unifying the several registration phases into `Register` which would make your particular error less likely

**Comment by knelli2** [comment]
@kidder Just found another instance of where unifying all registration phases into one `Register` phase is necessary on develop. Inside `InterpolationTarget.hpp`, it uses the `Register` phase to do some registration. In the GHWithHorizon executable we also use the `Register` phase. However, in the BBH executable we use the `RegisterWithObserver` phase. We use `InterpolationTarget` in both of these executables to do the exact same thing; register elements with the observer. So for the BBH exec, this causes the `Register` phase inside `InterpolationTarget` to never be run. I found this while doing GH+CCE.

**Comment by kidder** [comment]
@knelli2 okay I will put in a PR to unify the registration phases.  
I will also (attempt to) print out a warning if your phase action list contains phases not in the `default_phase_order`
